### ISSUE TITLE
Getting TypeError even when the object type is correct

### ISSUE THREAD

**Issue by kushubham9** [sylvainpolletvillard/ObjectModel#67]
Hi,
I am getting TypeError when trying to initialize a particular object model.
My classes and case is something like this.

class A extends Model.ObjectModel({
  id: [Number],
  refValue: [B],
 })

Class B extends Model.ObjectModel({
 c: [Number],
 d: [Number]
})

/// Now when I set a new instance of B,
let bObj = new B({c: 1, d:1 })
let aObj = new A({id: 5, refValue: bObj}) // This throws a TypeError.. expecting blah blah got Object [object Object]

This is strange to me as it working fine, but all of a sudden I started getting this error.

**Comment by sylvainpolletvillard** [comment]
Could you post the exact code you are trying ? The one you posted is full of syntax errors. I tried to fix it and it is working fine with latest version of ObjectModel:
```
class B extends ObjectModel({
c: [Number],
d: [Number]
}){}

class A extends ObjectModel({
id: [Number],
refValue: [B],
}){}

/// Now when I set a new instance of B,
let bObj = new B({c: 1, d:1 })
let aObj = new A({id: 5, refValue: bObj})
```

**Comment by kushubham9** [comment]
Above, I just posted a sample snippet to explain my case.

My Object init call:
![image](https://user-images.githubusercontent.com/16320695/40441356-6aa89e22-5ede-11e8-8144-3d5cf6d14010.png)

The object structure:
![image](https://user-images.githubusercontent.com/16320695/40441318-5071832a-5ede-11e8-9b2e-4616fc1c1200.png)

Error:
```
expecting pushMessage to be {
	pnsId: [Number],
	pushRequestId: [Number],
	isPersonalised: [Boolean],
	messageParams: [{
			messageId: [Number],
			title: [String],
			body: [String],
			url: [String],
			image: {
				small: [String],
				big: [String] 
			},
			timeToLive: [Number],
			autoHideNotificationFlag: [Number],
			autoHideNotificationSeconds: [Number],
			ctaButtons: [Array of {
					label: [String],
					url: [String] 
				}] 
		}],
	notificationTag: [String] 
}, got Object [object Object]
```

**Comment by sylvainpolletvillard** [comment]
It is very hard to help you based on screenshots. Could you find a simple and reproducible example ?

The last version is 3.4.5, released 28 days ago. You can try to downgrade to previous versions like 3.5.0 and see if that helps.

**Comment by kushubham9** [comment]
This is strange but when I change the object model to the following, it works:
```
class PushRequestEntity extends Model.ObjectModel({
  pushRequestId: [Number],
  customerId: [Number],
  target: [String],
  targetMetaData: [String],
  snapshot: [String],
  apiToken: [String],
  firstNotificationFlag: [Boolean],
  inProgress: [Boolean],
  scheduledNotificationId: [Number],
  pushMessage: PushMessage, // I removed the square brackets here.
  createdAt: [Date],
  modifiedAt: [Date]
})
```

Earlier it was:
  pushMessage: [PushMessage],


Any idea why is this happening with me? Isn't it so, if I remove the square brackets I make the field mandatory??

**Comment by kushubham9** [comment]
Ok I got it. So this was the issue.

In the 'PushRequestEntity' object I had a field 'pushRequestId' which is type Integer. I was setting the value of this field with a custom setter method

```
/**
   * @param prId {Number}
   */
  setPushRequestId (prId) {
    this.pushRequestId = prId
  }
```



```
let x = new PushRequestEntity({})
x.setPushRequestId("32") // Here I was passing a string by mistake

^ It gave no error when invoked
```


But when I pass this object 'x' as a parameter to another object instance (in the constructor), it was throwing a validation error.

^^^ When I use a setter for an object, why is that not validated on runtime??

**Comment by sylvainpolletvillard** [comment]
> When I use a setter for an object, why is that not validated on runtime??

It does:

![image](https://user-images.githubusercontent.com/566536/40445544-ac38beba-5ecc-11e8-9a9e-38d3273cc434.png)

 There must be something wrong with your code

**Comment by sylvainpolletvillard** [comment]
It is a limitation of the `class extends` pattern. With this pattern, the class is not a model, it **inherits** from a model. So if you add or modify a property that is owned by your class, and not the parent model, you won't get type-checking. 

In your setter, `this` refers to the class instance and not the parent model instance. In the parent model, the value is still undefined (based on the empty object in the constructor). And `undefined` is a valid value according to the model definition.

Code example :

![image](https://user-images.githubusercontent.com/566536/40446029-237da5b6-5ece-11e8-952e-9677728dc8cc.png)

**Comment by sylvainpolletvillard** [comment]
Maybe I found a way to get around this limitation. I need to do some testing though. Stay tuned

**Comment by sylvainpolletvillard** [comment]
I just published v3.4.6 which should solves this issue. Can you try again with this version ?

**Comment by kushubham9** [comment]
It works. Thanks a lot.
